ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -22,6 +21,12 @@ class UpdateLedger blk => ApplyTx blk where
   -- also other kinds of things such as update proposals, delegations, etc.
   data family GenTx blk :: *
 
+  -- | A generalized transaction, 'GenTx', identifier.
+  data family GenTxId blk :: *
+
+  -- | Given a 'GenTx', compute its 'GenTxId'.
+  getGenTxId :: GenTx blk -> GenTxId blk
```

Comment:
Not sure whether anyone disagrees with adding this to the `ApplyTx` typeclass. I required this kind of functionality in [`localTxSubmissionServer`](https://github.com/input-output-hk/ouroboros-network/pull/678/files#diff-81e1b5d57b8f896bf0eb321dbe640688R34).

Reply:
Yes it makes perfect sense to be in the same class as the `GenTx` and `GenTxId` types.

Naming suggestion: make it sound expensive rather than like a field accessor, so it is clear why we cache it.

The alternative design is to require that the id is cached and that the function just returns the cached value. That'd allow a design of the mempool API where we only need to return txs, not txids and txs, since accessing one from the other would be cheap.

Either is fine, but we should document in this class which approach instances are expected to take so that the costs are clear.